### PR TITLE
Base64 decoder rust/v15

### PR DESCRIPTION
**Pull request by inashivb**
Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7065
- https://redmine.openinfosecfoundation.org/issues/7058
- https://redmine.openinfosecfoundation.org/issues/6280

Previous PR: https://github.com/OISF/suricata/pull/11734

Changes since v14:
- rebased on top of latest master to get green CI

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.dcerpc_tcp | 40 | 42 | 105.0% | 

Pipeline 22646

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
  

Pipeline 22677

File: src/detect-base64-decode.c [hunk 1]
```diff
@@ -89,17 +89,18 @@ int DetectBase64DecodeDoMatch(DetectEngineThreadCtx *det_ctx, const Signature *s
 
     decode_len = MIN(payload_len, data->bytes);
 
+    DEBUG_VALIDATE_BUG_ON(decode_len < 0);
```
Comment (catenacyber) [inline src/detect-base64-decode.c]: Which could be negative ? payload_len or data->bytes ?

File: rust/src/ffi/base64.rs [hunk 2]
```diff
@@ -27,6 +28,90 @@ pub enum Base64ReturnCode {
...
+
+/// Base64 decode a buffer.
+///
+/// This method exposes the Rust base64 decoder to C and should not be called from
+/// Rust code.
+///
+/// It allows decoding in the modes described by ``Base64Mode`` enum.
+#[no_mangle]
+pub unsafe extern "C" fn Base64Decode(
+    input: *const u8, len: usize, max_decoded: u32, mode: Base64Mode, output: *mut u8) -> u32 {
```
Comment (catenacyber) [inline rust/src/ffi/base64.rs:82]: What is the purpose of `max_decoded` argument ?

File: rust/src/mime/smtp.rs [hunk 3]
```diff
@@ -668,14 +546,17 @@ fn mime_smtp_parse_line(
                             if i.len() > MAX_ENC_LINE_LEN {
                                 warnings |= MIME_ANOM_LONG_ENC_LINE;
                             }
-                            if let Ok(dec) = mime_base64_decode(decoder, i) {
+                            let dec_size = ((decoder.nb as usize + i.len()) * 3) / 4;
```
Comment (catenacyber) [inline rust/src/mime/smtp.rs:549]: Is this not `Base64DecodeBufferSize` ?

File: rust/src/mime/smtp.rs [hunk 4]
```diff
@@ -216,7 +198,7 @@ fn mime_smtp_process_headers(ctx: &mut MimeStateSMTP) -> (u32, bool) {
         } else if mime::slice_equals_lowercase(&h.name, b"content-transfer-encoding") {
             if mime::slice_equals_lowercase(&h.value, b"base64") {
                 ctx.encoding = MimeSmtpEncoding::Base64;
-                ctx.decoder = Some(MimeBase64Decoder::new());
+                ctx.decoder = Some(base64::Decoder::new());
```
Comment (catenacyber) [inline rust/src/mime/smtp.rs:201]: @jasonish do we risk naming conflicts with rust for using `base64` ?

**Comment by catenacyber** [comment]
This looks awesome :-)

File: src/detect-base64-decode.c (hunk 1 above)
Comment (inashivb) [reply src/detect-base64-decode.c]: `payload_len`, I believe as it is being calculated and subtracted from right above this. I tried to make this better in the following commit btw https://github.com/OISF/suricata/pull/11783/commits/ee6263eb328c017f7dc408b1d7957f0e460188b9 Lmk wdyt?

File: rust/src/ffi/base64.rs (hunk 2 above)
Comment (inashivb) [reply rust/src/ffi/base64.rs:82]: This is used particularly by `base64.decode` to indicate the maximum number of bytes we want to decode. I suppose we should consider MIN(encoded_len*3/4, max_decoded) as the actual number of bytes that should be decoded. Looking at the code, I'm not doing that so this needs fixing. Thanks a lot! 🙇🏽‍♀️

File: rust/src/mime/smtp.rs (hunk 3 above)
Comment (inashivb) [reply rust/src/mime/smtp.rs:549]: indeed

File: rust/src/mime/smtp.rs (hunk 4 above)
Comment (inashivb) [reply rust/src/mime/smtp.rs:201]: We don't. I was worried about the same and Jason suggested it be done this way. There was already `ffi/base64.rs` which uses the rust `base64` crate. :)

File: src/detect-base64-decode.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-base64-decode.c]: Indeed https://github.com/OISF/suricata/commit/ee6263eb328c017f7dc408b1d7957f0e460188b9 looks good

File: rust/src/ffi/base64.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/ffi/base64.rs:82]: When is not `max_decoded` equal to `Base64DecodeBufferSize(len)` ?

Comment (catenacyber) [reply rust/src/ffi/base64.rs:82]: (and my point was rather to have a comment explaining this argument ;-) )

Comment (inashivb) [reply rust/src/ffi/base64.rs:82]: > When is not max_decoded equal to Base64DecodeBufferSize(len) ?

Turns out it is always much bigger. So, this arg can be safely removed and I don't need to do any modifications.

**Pull request closed without merging**